File: packages/spectacular-llms/src/openai/debug.rs
use crate::debug as provider_debug;
use crate::{LlmDebugLogger, ProviderError};

const TARGET: &str = "openai";

/// Writes an OpenAI debug event without failing the provider call.
pub(crate) fn log_event(logger: &LlmDebugLogger, event: &str, fields: serde_json::Value) {
    provider_debug::log_event(TARGET, logger, event, fields);
}

/// Writes a raw JSON OpenAI debug event without failing the provider call.
pub(crate) fn log_raw_json(logger: &LlmDebugLogger, event: &str, raw_json: serde_json::Value) {
    provider_debug::log_raw_json(TARGET, logger, event, raw_json);
}

/// Writes a raw text OpenAI debug event without failing the provider call.
pub(crate) fn log_raw_text(logger: &LlmDebugLogger, event: &str, raw_text: &str) {
    provider_debug::log_raw_text(TARGET, logger, event, raw_text);
}

/// Writes an OpenAI error debug event without failing the provider call.
pub(crate) fn log_error(logger: &LlmDebugLogger, event: &str, error: &ProviderError) {
    provider_debug::log_error(TARGET, logger, event, error);
}

File: packages/llms/src/lib/providers/openai/debug.ts
import { ProviderErrorObject } from '../../classes/provider-error.js';
import type { LlmDebugLogger } from '../../debug.js';
import type {
  JsonValue,
  ProviderFinished,
  ProviderId,
  ProviderRequest,
} from '../../types/provider.js';
import { diagnosticExcerpt } from '../../utils/diagnostics.js';
import { parseStructuredOutput } from '../common.js';

export const parseStructuredOutputWithDebug = async <Output = JsonValue>(
  logger: LlmDebugLogger | undefined,
  debugProvider: ProviderId,
  request: ProviderRequest<Output>,
  finish: ProviderFinished,
  source: string,
  rejectNonStructured = true,
): Promise<ProviderFinished<Output>> => {
  const sensitiveOutput = request.flags?.sensitiveOutput === true;
  await logger?.log({
    provider: debugProvider,
    target: 'responses',
    event: 'response.finish',
    fields: { source, finish: finishDebugFields(finish, sensitiveOutput) },
  });

  try {
    return parseStructuredOutput(
      'openai',
      request,
      finish,
      rejectNonStructured,
    );
  } catch (error) {
    await logger?.log({
      provider: debugProvider,
      target: 'responses',
      event: 'structured_output.error',
      fields: {
        source,
        finish: finishDebugFields(finish, sensitiveOutput),
        error: errorDebugFields(error, sensitiveOutput),
      },
    });
    throw error;
  }
};

const finishDebugFields = (
  finish: ProviderFinished,
  sensitiveOutput: boolean,
): Record<string, unknown> => ({
  finishReason: finish.finishReason,
  textLength: finish.text.length,
  ...(sensitiveOutput
    ? {}
    : { textExcerpt: diagnosticExcerpt(finish.text, 512) }),
  refusalPresent: finish.refusal !== undefined,
  reasoningPresent: finish.reasoning?.text !== undefined,
  toolCallCount: finish.toolCalls.length,
  toolCalls: finish.toolCalls.map((call) => ({
    id: call.id,
    name: call.name,
    argumentsLength: call.arguments.length,
  })),
  usage: finish.usage,
});

const errorDebugFields = (
  error: unknown,
  sensitiveOutput: boolean,
): Record<string, unknown> => {
  if (error instanceof ProviderErrorObject) {
    return {
      name: error.name,
      code: error.data.code,
      message: error.data.message,
      ...(sensitiveOutput
        ? {}
        : {
            diagnostic: error.data.diagnostic,
            cause: causeDebugFields(error),
          }),
    };
  }

  if (error instanceof Error) {
    return {
      name: error.name,
      message: error.message,
      ...(sensitiveOutput ? {} : { cause: causeDebugFields(error) }),
    };
  }

  return sensitiveOutput
    ? { value: 'Sensitive provider error.' }
    : { value: String(error) };
};

const causeDebugFields = (
  error: Error & { readonly cause?: unknown },
): Record<string, unknown> | undefined => {
  const cause = error.cause;

  if (cause instanceof Error) {
    return { name: cause.name, message: cause.message };
  }

  return cause === undefined ? undefined : { value: String(cause) };
};
